File: DSaA/2024-2025/contests/564591/A. Длиннейший путь/main.py
n, m = map(int, input().split())

edges = []
for _ in range(m):
    u, v = map(int, input().split())
    edges.append([u, v])

adj = [[] for _ in range(n + 1)]


def tsort(v, visited, order):
    visited[v] = True
    for u in adj[v]:
        if not visited[u]:
            tsort(u, visited, order)
    order.append(v)


for edge in edges:
    adj[edge[0]].append(edge[1])

visited = [False] * (n + 1)
order = []
for i in range(1, n + 1):
    if not visited[i]:
        tsort(i, visited, order)

order.reverse()

dist = [0] * (n + 1)
for u in order:
    for v in adj[u]:
        dist[v] = max(dist[v], dist[u] + 1)

print(max(dist))

File: DSaA/2024-2025/contests/564591/A. Длиннейший путь/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	line, _ := reader.ReadString('\n')
	line = strings.TrimSpace(line)
	parts := strings.Split(line, " ")

	n, _ := strconv.Atoi(parts[0])
	m, _ := strconv.Atoi(parts[1])

	edges := make([][]int, m)
	for i := 0; i < m; i++ {
		line, _ := reader.ReadString('\n')
		line = strings.TrimSpace(line)
		parts := strings.Split(line, " ")
		u, _ := strconv.Atoi(parts[0])
		v, _ := strconv.Atoi(parts[1])
		edges[i] = []int{u, v}
	}

	adj := make([][]int, n+1)
	for _, edge := range edges {
		u, v := edge[0], edge[1]
		adj[u] = append(adj[u], v)
	}
	visited := make([]bool, n+1)
	order := []int{}

	var tsort func(v int)
	tsort = func(v int) {
		visited[v] = true
		for _, u := range adj[v] {
			if !visited[u] {
				tsort(u)
			}
		}
		order = append(order, v)
	}

	for i := 1; i <= n; i++ {
		if !visited[i] {
			tsort(i)
		}
	}

	for i, j := 0, len(order)-1; i < j; i, j = i+1, j-1 {
		order[i], order[j] = order[j], order[i]
	}

	dist := make([]int, n+1)
	for _, u := range order {
		for _, v := range adj[u] {
			if dist[v] < dist[u]+1 {
				dist[v] = dist[u] + 1
			}
		}
	}

	maxDist := 0
	for i := 1; i <= n; i++ {
		if dist[i] > maxDist {
			maxDist = dist[i]
		}
	}

	fmt.Println(maxDist)
}
